rfcs/text/0033/io.yml

```diff
@@ -0,0 +1,43 @@
+- name: io
+  title: IO
+  group: 2
+  short: Fields describing a batch of input or output data produced by a process.
+  description: >
+    Fields describing a batch of input or output (IO) data produced by a process.
+    These fields should only appear on the top level process object when event.action is 'io'
+  type: group
+  reusable:
+    top_level: false
+    expected:
+      - process
+
+  fields:
+
+    - name: data
+      level: extended
+      type: object
+      short: An array of timestamped raw input/output lines encoded as UTF-8.
+      description: >
+        An array of timestamped raw input/output lines encoded as UTF-8 characters.
+
+        Each event can contain up to <batch size> (configurable) lines of data. With a maximum line length of <X> (needs discussion, regex?, scrubbing?).
+
+      normalize: array
+
+    - name: data.time
+      level: extended
+      type: date
+      short: Timestamp for the line of IO
+      description: >
+        Timestamp for the line of IO
+
+        Timestamps are tracked per line, so that they can be replayed in realtime using libraries like xtermjs.org.
+
+    - name: data.value
+      level: extended
+      type: text
```

Comment:
If using `text` types, I recommend evaluating [`match_only_text`](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#match-only-text-field-type) over `text` for the better space efficiency. 

However, using `text` or `match_only_text` will use the `standard` analyzer, which indexes using grammar-based tokenization and is case-insensitive. Depending on how users interact with these fields, the search results may be unexpected.

The standard analyzer will remove the special characters when indexing the value `#!/bin/bash`, leaving only `bin` and `bash` tokens:

```
POST /_analyze
{
  "text": "#!/bin/bash"
}

{
  "tokens" : [
    {
      "token" : "bin",
      "start_offset" : 3,
      "end_offset" : 6,
      "type" : "<ALPHANUM>",
      "position" : 0
    },
    {
      "token" : "bash",
      "start_offset" : 7,
      "end_offset" : 11,
      "type" : "<ALPHANUM>",
      "position" : 1
    }
  ]
}
```

The [`wildcard` type](https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html#wildcard-field-type) is an alternative used a bit in ECS that won't strip special characters when indexing and is also case-sensitive.

Reply:
Ah good catch. I think it would be strange if we couldn't match on certain special characters, especially forward and back slashes. Though there are ascii control characters in this IO data which might benefit from being excluded in these searches, as they can appear between words. (an issue @m-sample and I have chatted about). 

It probably makes sense to use wildcard type. Any thoughts @m-sample ?

Reply:
Back from vacation now, looking at this today.

Reply:
updated to use keyword type.

Reply:
Not sure why I updated to keyword type, I think I had it in my head that was what was mentioned above. Anyways this came up again today. Took a peek at https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html#wildcard-field-type

It seems reasonable we'd want to use wildcard type, as I'm sure we'll have events with > 32kb of io data. And output data will most always be searched with a wildcard. Though, since process.io.data is an array of objects, each having a value that is a 'wildcard' type. Each line would need to exceed 32kb for it to be worth it? Sounds like wildcard type comes at a cost to storage and speed though.